private-templates-service/adaptivecards-templating-service/src/storageproviders/MongoDBProvider.ts

```diff
@@ -0,0 +1,208 @@
+import mongoose from "mongoose";
+import { StorageProvider } from "./IStorageProvider";
+import { IUser, ITemplate, JSONResponse } from "../models/models";
+import { IUserModel, UserSchema } from "../models/mongo/UserModel";
+import { ITemplateModel, TemplateSchema } from "../models/mongo/TemplateModel";
+import { Mongo } from "../util/mongoutils";
+
+const USERS_COLLECTION_NAME_SINGULAR: string = "User";
+const TEMPLATES_COLLECTION_NAME_SINGULAR: string = "Template";
+
+export class MongoDBProvider implements StorageProvider {
+  Template!: mongoose.Model<ITemplateModel>;
+  User!: mongoose.Model<IUserModel>;
+  db!: mongoose.Connection;
+  connectionString: string;
+  options: any;
+  async getUser(query: Partial<IUser>): Promise<JSONResponse<IUser[]>> {
+    return await this.User.find(query)
+      .then(users => {
+        if (users.length) {
+          return Promise.resolve({ success: true, result: users });
+        }
+        return Promise.resolve({
+          success: false,
+          result: [],
+          errorMessage: "No users found matching given criteria"
+        });
+      })
+      .catch(e => {
+        return Promise.resolve({ success: false, errorMessage: e });
+      });
+  }
+  async getTemplate(query: Partial<ITemplate>): Promise<JSONResponse<ITemplate[]>> {
+    return await this.Template.find(query)
+      .then(templates => {
+        if (templates.length) {
+          return Promise.resolve({
+            success: true,
+            result: templates
+          });
+        }
+        return Promise.resolve({
+          success: false,
+          result: [],
+          errorMessage: "No templates found matching given criteria"
+        });
+      })
+      .catch(e => {
+        return Promise.resolve({ success: false, errorMessage: e });
+      });
+  }
+  // Updates Only one user
+  async updateUser(query: Partial<IUser>, updateQuery: Partial<IUser>): Promise<JSONResponse<Number>> {
```

Comment:
Simpler to use Boolean instead of Number? (Or if it's just used to indicate success, don't send a result, just use the 'success' flag).

Reply:
success flag is gonna be sent, but I need to set the type of the **JSONResponse**, boolean would be redundant but with number I can just send 0 as # of updated users :)

Reply:
- You've hard-coded the return numbers 0 and 1 (effectively the same as a boolean) so you're not providing any meaningful result information.
-The function name already indicates that one user will be updated, as it's called 'updateUser'.

I don't think you need to send a result in this case.

Reply:
You're right :)
Fixed :)
Though when you will look at `getUser|template` functions, you'll see that I will send the result in case of failure as `[]`. The reason for that is some weird type deduction by TS which is not working properly so it fails to compile. Screenshot attached :) 
![image](https://user-images.githubusercontent.com/21356912/74184373-89dc1d80-4bfb-11ea-9391-4c0f6fec0c96.png)

Reply:
I think you can fix this by just specifying the type of 'users' (i.e. `.then((users: IUser[]) => ...` )?

I think for consistency we should decide now whether the 'result' of JSONResponse should be optional or not. I'm good with either, but we should make everything the same. If there is no meaningful info to return, and _result_ is optional, don't include it. In scenarios like this _updateUsers_ function, and if _result_ is non-optional, use Boolean.